lang/java/avro/src/main/java/org/apache/avro/SchemaNormalization.java

```diff
@@ -160,14 +200,73 @@ private static Appendable build(Map<String, String> env, Schema s, Appendable o)
           else
             firstTime = false;
           o.append("{\"name\":\"").append(f.name()).append("\"");
-          build(env, f.schema(), o.append(",\"type\":")).append("}");
+          build(env, f.schema(), o.append(",\"type\":"), ps, aps);
+          if (!ps)
+            setFieldProps(o, f, aps); // if standard canonical form then add reserved properties
+          o.append("}");
         }
         o.append("]");
       }
+      if (!ps) {
+        setComplexProps(o, s);
+        setSimpleProps(o, s.getObjectProps(), aps);
+      } // adding the reserved property if not parser canonical schema
       return o.append("}");
     }
   }
 
+  private static Appendable writeLogicalType(Schema s, LogicalType lt, Appendable o, LinkedHashSet<String> aps)
+      throws IOException {
+    o.append("{\"type\":\"").append(s.getType().getName()).append("\"");
+    // adding the logical property
+    setLogicalProps(o, lt);
+    // adding the reserved property
+    setSimpleProps(o, s.getObjectProps(), aps);
+    return o.append("}");
+  }
+
+  private static void setLogicalProps(Appendable o, LogicalType lt) throws IOException {
+    o.append(",\"").append(LogicalType.LOGICAL_TYPE_PROP).append("\":\"").append(lt.getName()).append("\"");
+    if (lt.getName().equals("decimal")) {
+      LogicalTypes.Decimal dlt = (LogicalTypes.Decimal) lt;
+      o.append(",\"precision\":").append(Integer.toString(dlt.getPrecision()));
+      if (dlt.getScale() != 0)
+        o.append(",\"scale\":").append(Integer.toString(dlt.getScale()));
+    }
+  }
+
+  private static void setSimpleProps(Appendable o, Map<String, Object> schemaProps, LinkedHashSet<String> aps)
+      throws IOException {
+    for (String propKey : aps) {
+      if (schemaProps.containsKey(propKey)) {
+        String propValue = JacksonUtils.toJsonNode(schemaProps.get(propKey)).toString();
+        o.append(",\"").append(propKey).append("\":").append(propValue);
+      }
+    }
+  }
+
+  private static void setComplexProps(Appendable o, Schema s) throws IOException {
+    if (s.getDoc() != null && !s.getDoc().isEmpty())
+      o.append(",\"doc\":\"").append(s.getDoc()).append("\"");
```

Comment:
This should be `o.append(",\"doc\":").append(toJsonNode(s.getDoc()).toString());`